src/mds/MDCache.cc

```diff
@@ -12781,15 +12781,25 @@ int MDCache::dump_cache(std::string_view fn, Formatter *f)
     return 1;
   };
 
+  auto start = mono_clock::now();
+  int64_t count = 0;
   for (auto &p : inode_map) {
     r = dump_func(p.second);
     if (r < 0)
       goto out;
+    if (!(++count % 1000) &&
+	timeout > 0 &&
+	std::chrono::duration<double>(mono_clock::now() - start).count() > timeout)
+      goto out;
   }
   for (auto &p : snap_inode_map) {
     r = dump_func(p.second);
     if (r < 0)
       goto out;
+    if (!(++count % 1000) &&
+		timeout > 0 &&
+	std::chrono::duration<double>(mono_clock::now() - start).count() > timeout)
+      goto out;
```

Comment:
After closing the `"inodes"` section,  how about we add some information to the json to indicate the `inodes` array is incomplete due to the timeout. Otherwise the caller has no way to differentiate the two results.